core/utils/historical-data/src/linear.rs

```diff
@@ -0,0 +1,577 @@
+	}
+
+	/// Commit prospective changes to state.
+	pub fn commit_prospective(&mut self) {
+		self.commit = self.history.len();
+		self.history.push(TransactionState::Pending);
+		for i in 0..self.history.len() - 1 {
+			self.previous_transaction[i] = self.commit;
+		}
+		self.previous_transaction.push(self.commit);
+	}
+
+	/// Create a new transactional layer.
+	pub fn start_transaction(&mut self) {
+		self.previous_transaction.push(self.history.len());
+		self.history.push(TransactionState::TxPending);
+	}
+
+	/// Discard a transactional layer.
+	/// A transaction is always running (history always end with pending).
+	pub fn discard_transaction(&mut self) {
+		let mut i = self.history.len();
+		while i > self.commit {
+			i -= 1;
+			match self.history[i] {
+				TransactionState::Dropped => (),
+				TransactionState::Pending => {
+					self.history[i] = TransactionState::Dropped;
+				},
+				TransactionState::TxPending => {
+					self.history[i] = TransactionState::Dropped;
+					break;
+				},
+			}
+		}
+		self.history.push(TransactionState::Pending);
+		self.previous_transaction.truncate(i);
+		let previous = self.previous_transaction.last()
+			.cloned().unwrap_or(self.commit);
+		self.previous_transaction.resize(self.history.len(), previous);
+	}
+
+	/// Commit a transactional layer.
+	pub fn commit_transaction(&mut self) {
+		let mut i = self.history.len();
+		while i > self.commit {
+			i -= 1;
+			match self.history[i] {
+				TransactionState::Pending
+				| TransactionState::Dropped => (),
+				TransactionState::TxPending => {
+					self.history[i] = TransactionState::Pending;
+					break;
+				},
+			}
+		}
+		self.history.push(TransactionState::Pending);
+		self.previous_transaction.truncate(i);
+		let previous = self.previous_transaction.last()
+			.cloned().unwrap_or(self.commit);
+		self.previous_transaction.resize(self.history.len(), previous);
+	}
+
+}
+
+#[inline]
+/// Get previous index of pending state.
+///
+/// Used to say if it is possible to drop a committed transaction
+/// state value.
+/// Committed index is seen as a transaction state.
+pub fn find_previous_tx_start(states: &States, from: usize) -> usize {
+	states.previous_transaction[from]
+}
+
+impl<V> History<V> {
+	/// Set a value, it uses a state history as parameter.
+	/// This method uses `get_mut` and does remove pending
+	/// dropped value.
+	pub fn set(&mut self, states: &States, value: V) {
+		if let Some(v) = self.get_mut(states) {
+			if v.index == states.history.len() - 1 {
+				*v.value = value;
+				return;
+			}
+		}
+		self.push_unchecked(HistoricalValue {
+			value,
+			index: states.history.len() - 1,
+		});
+	}
+
+	/// Access to latest pending value (non dropped state).
+	/// When possible please prefer `get_mut` as it can free
+	/// some memory.
+	pub fn get(&self, states: &[TransactionState]) -> Option<&V> {
+		// index is never 0,
+		let mut index = self.len();
+		if index == 0 {
+			return None;
+		}
```

Comment:
It was to indicate that a 'states' shall always have at least a pending layer, but this comment do not make sense in this context since we do manage the 0 case, I am removing it (it was also somewhere else).